core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1907,5 +1916,23 @@ private boolean isAddedSnapshot(long snapshotId) {
     private <U extends MetadataUpdate> Stream<U> changes(Class<U> updateClass) {
       return changes.stream().filter(updateClass::isInstance).map(updateClass::cast);
     }
+
+    /**
+     * Parquet is currently the only format compatible with v3 due to missing support for features
+     * like default values, type support, etc. in other formats which make creating or updating to
+     * v3 unsafe. Supported formats should be expanded as support is implemented.
+     */
+    private static void validateFileFormatCompatibility(
+        int formatVersion, Map<String, String> properties) {
+      FileFormat targetFormat =
+          FileFormat.fromString(
+              properties.getOrDefault(DEFAULT_FILE_FORMAT, DEFAULT_FILE_FORMAT_DEFAULT));
+
+      Preconditions.checkArgument(
+          V3_SUPPORTED_FILE_FORMATS.contains(targetFormat) || formatVersion <= 2,
```

Comment:
this is probably fine for now. 

but it is tied specifically to v3. if we are adding v4 in the future, this is another another we need to keep up. we probably need find some more elegant way to manage all those version specific validations.

Reply:
Agree. Maybe the reverse is more cleaner, like maximum supported table format version in `FileFormat.java` for each format[1]. So, here validation can be generic for all file format and table format.  

[1]
Parquet - 3
avro - 2
ORC - 2
PUFFIN- 2

Reply:
I considered this as well, but it's a little more complicated than just having a max supported version.  For example, Avro is supported in 1,2 and 3 as a metadata format (we still write manifests using Avro), but currently only 1,2 for data format.  You also can't write formats other than puffin for deletes with 3 and beyond.

I'll think about this and see if there's a clearer way to present this.

Reply:
I think this is a good approach for now. We don't need to build too much here since we intend to remove this and resolve the compatibility issues. If that doesn't happen then we actually need to address it in a different way, by working as a community to deprecate support and make this more permanent.

> PUFFIN- 2

Puffin should be v3 and beyond. We can't write Puffin files for DVs in v2.